subscription.go

```diff
@@ -505,31 +533,35 @@ func (sc *SubscriptionClient) Unsubscribe(id string) error {
 	return sc.protocol.Unsubscribe(sc.context, id)
 }
 
-// Run start websocket client and subscriptions. If this function is run with goroutine, it can be stopped after closed
+// Run start the WebSocket client and subscriptions.
+// If the client is running, recalling this function will restart all registered subscriptions
+// If this function is run with goroutine, it can be stopped after closed
 func (sc *SubscriptionClient) Run() error {
+
+	sc.reset()
 	if err := sc.init(); err != nil {
 		return fmt.Errorf("retry timeout. exiting...")
 	}
 
 	sc.setIsRunning(true)
 	go func() {
-		for atomic.LoadInt64(&sc.isRunning) > 0 {
+		for sc.getIsRunning() {
 			select {
-			case <-sc.context.Done():
+			case <-sc.context.GetContext().Done():
 				return
 			default:
 				if sc.context == nil || sc.context.GetWebsocketConn() == nil {
 					return
 				}
 
 				var message OperationMessage
-				if err := sc.context.ReadJSON(&message); err != nil {
+				if err := sc.context.GetWebsocketConn().ReadJSON(&message); err != nil {
 					// manual EOF check
 					if err == io.EOF || strings.Contains(err.Error(), "EOF") {
-						if err = sc.Reset(); err != nil {
+						if err = sc.Run(); err != nil {
```

Comment:
Calling Run in a recursive way, leaks goroutines, which increases complexity of this function and could crash the go runtime (in theory) due to an always increasing goroutine set. Have you considered replacing recursion with a loop?

Reply:
You're right. Updated